### PR TITLE
Add implementations for random_saturation

### PR DESCRIPTION
**Pull request by shashaka**
I have implemented the RandomSaturation image processing feature. Please let me know if any further corrections or improvements are required.

here is my [gist](https://colab.research.google.com/drive/147W37NYgFVeifRNN1m2pa2xNTxpg2cIB#scrollTo=dG01ookclJuO)

### CODE REVIEW

File: keras/src/layers/preprocessing/image_preprocessing/random_saturation.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+            images = data
+        images_shape = self.backend.shape(images)
+        rank = len(images_shape)
+        if rank == 3:
+            batch_size = 1
+        elif rank == 4:
+            batch_size = images_shape[0]
+        else:
+            raise ValueError(
+                "Expected the input image to be rank 3 or 4. Received "
```
Comment (fchollet) [inline keras/src/layers/preprocessing/image_preprocessing/random_saturation.py:56]: Add `:` after received

File: keras/src/layers/preprocessing/image_preprocessing/random_saturation.py [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+            maxval=self.factor[1],
+            seed=seed,
+        )
+        factor = factor / (1 - factor)
+        return {"factor": factor}
+
+    def transform_images(self, images, transformation=None, training=True):
+        if training:
+            adjust_factors = transformation["factor"]
+            adjust_factors = self.backend.cast(adjust_factors, images.dtype)
```
Comment (fchollet) [inline keras/src/layers/preprocessing/image_preprocessing/random_saturation.py:75]: This can be dangerous since `images` could be int, right? Better to convert all to `self.compute_dtype`

File: keras/src/layers/preprocessing/image_preprocessing/random_saturation.py [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+            adjust_factors = self.backend.numpy.expand_dims(adjust_factors, -1)
```
Comment (fchollet) [inline keras/src/layers/preprocessing/image_preprocessing/random_saturation.py:76]: Since can be done in a single reshape call

File: keras/src/layers/preprocessing/image_preprocessing/random_saturation.py [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+
+            images = self.backend.image.rgb_to_hsv(
+                images, data_format=self.data_format
+            )
+
+            if self.data_format == "channels_first":
+                s_channel = self.backend.numpy.multiply(
+                    images[:, 1, :, :], adjust_factors
+                )
+                s_channel = self.backend.numpy.clip(s_channel, 0.0, 1.0)
```
Comment (fchollet) [inline keras/src/layers/preprocessing/image_preprocessing/random_saturation.py:87]: You are assuming input values in [0, 1] but they could also be in [0, 255]. We should rely on `self.value_range`

File: keras/src/layers/preprocessing/image_preprocessing/random_saturation.py [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+    Example:
+    ```python
+    (images, labels), _ = keras.datasets.cifar10.load_data()
+    images = images.astype("float32") / 255.0
+    random_saturation = keras.layers.RandomSaturation(factor=0.2)
+    augmented_images = random_saturation(images)
+    ```
+    """
+
+    def __init__(self, factor, data_format=None, seed=None, **kwargs):
```
Comment (fchollet) [inline keras/src/layers/preprocessing/image_preprocessing/random_saturation.py:37]: Add `value_range` argument (see other layers, like `RandomBrightness`)

**Review by fchollet — COMMENTED** [review]
Thanks for the PR!

File: keras/src/layers/preprocessing/image_preprocessing/random_saturation.py (hunk 1 above)
Comment (shashaka) [reply keras/src/layers/preprocessing/image_preprocessing/random_saturation.py]: I've added `:`

File: keras/src/layers/preprocessing/image_preprocessing/random_saturation.py (hunk 2 above)
Comment (shashaka) [reply keras/src/layers/preprocessing/image_preprocessing/random_saturation.py]: I've changed it to self.compute_dtype

File: keras/src/layers/preprocessing/image_preprocessing/random_saturation.py (hunk 4 above)
Comment (shashaka) [reply keras/src/layers/preprocessing/image_preprocessing/random_saturation.py]: I've changed it to use self.value_range.

File: keras/src/layers/preprocessing/image_preprocessing/random_saturation.py [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+        )
+        factor = factor / (1 - factor)
+        return {"factor": factor}
+
+    def transform_images(self, images, transformation=None, training=True):
+        if training:
+            images = self._apply_random_saturation(images, transformation)
+        return images
+
+    def _apply_random_saturation(self, images, transformation):
```
Comment (fchollet) [inline keras/src/layers/preprocessing/image_preprocessing/random_saturation.py:106]: Please inline this function in `transform_images`, there's no need to make it a separate method (it is only used 1x)

**Review by fchollet — COMMENTED** [review]
The update looks good!

File: keras/src/layers/preprocessing/image_preprocessing/random_saturation.py (hunk 6 above)
Comment (shashaka) [reply keras/src/layers/preprocessing/image_preprocessing/random_saturation.py]: I have refactored this function to be inline.

**Review by fchollet — APPROVED** [review]
LGTM, thanks!
